Make it so that uacomment[] are correctly reported in subver string

sickpig opened the pull request:
Superseding #1827 and fix #1061

sickpig committed:
Make it so that uacomment[] are correctly reported in subver string.

This is a fix for issue #1601. Namely "uaccoments" parameters were
overwritten by EB/AD settings. These change are always appending
"uaccoments" to subver string in the init step and also at run time
when EB and/or AD are changed.

If the resulting subver is longer than 256 chars (per BIP 014), in the
init step we are going to emit a warning and exit. If this situation triggers
at run time we just trim the subver until it met the 256 chars limit.

sickpig committed:
Test subver string for "uaccoment" and length

ptschip reviewed (approved):
Tested and works...thanks!

ptschip committed:
Add a 32bit identifier to the uacomment string

This will help us know whether we still need to support 32bit
releases on mainnet.

Justaphf commented on src/qt/clientmodel.cpp:
```diff
@@ -142,7 +142,10 @@ OptionsModel *ClientModel::getOptionsModel() { return optionsModel; }
 PeerTableModel *ClientModel::getPeerTableModel() { return peerTableModel; }
 BanTableModel *ClientModel::getBanTableModel() { return banTableModel; }
 QString ClientModel::formatFullVersion() const { return QString::fromStdString(FormatFullVersion()); }
-QString ClientModel::formatSubVersion() const { return QString::fromStdString(strSubVersion); }
```
I think with this change strSubVersion can be removed from global state (net.h/.cpp) and just made a local variable in init.cpp, it isn’t used anywhere else.

sickpig replied:
fixed

Justaphf commented on src/init.cpp:
```diff
@@ -1337,14 +1337,16 @@ bool AppInit2(Config &config, thread_group &threadGroup)
+
+    strSubVersion = FormatSubVersion(CLIENT_NAME, CLIENT_VERSION, BUComments);
+
     if (strSubVersion.size() > MAX_SUBVERSION_LENGTH)
```
With truncation being done in FormatSubVersion now, this check can never fail.
This also represents a small loss in functionality because now with truncation the user will never receive feedback that their uacomment(s) as specified exceed the maximum length, they will just silently be clipped to fit.  I don’t know if it should still constitute a failure to start since the string length is variable at runtime if the user changes EB/AD settings, but a warning would still be good when clipping occurs.

sickpig replied:
fixed

Justaphf commented:
@sickpig thanks for working on this. I left 2 comments based solely on code review. I’ll do implementation testing tomorrow when I’m back from travel.

Justaphf reviewed (changes requested):
@sickpig Other than the 2 items I already mentioned, implementation testing looks good.

I did notice a couple of other minor issues that are out of scope for this ticket, just noting them here since I ran into them while testing.
1. Really long user agent strings cause the debug window to force itself to a really wide size.  Solution is probably to make the QT string field either auto-wrap or truncate instead of increasing the minimum size of the dialog.  Only affects the user who's set the long custom user agent string.
2. When changing the EB/AD at runtime via RPC `set`, the user agent string displayed in the debug window is not updated.  This string will only ever show what was configured at the time that the dialog is first displayed.  This does not affect the RPC results for `getnetworkinfo` or the subversion string sent to peers that are connected after the EB/AD change is made.

NOTE: Requested changes only for the two items in code locations, not the out of scope items listed above.

sickpig commented:
@Justaphf thanks for the review. Will address points raised asap.

sickpig committed:
Make strSubVersion local to init.cpp

sickpig committed:
Warn about excessive subver length in FormatSubVersion()

Moving the check of lenght and trimming in FormatSubVersion(),
make it so that the check for lenght in init.cpp can never fail and
InitError() never trigger.

Wiithout this change useisr would never have been notified about setting
a uacomment param that produce a too long subver string.

sickpig commented:
@Justaphf fixed the two point you raised in your code review. Will work on the other 2 points you raised (long subver auto wrap, updating subcer in qt debug window) if nobody beats to me.

Justaphf commented on src/clientversion.cpp:
```diff
@@ -116,17 +116,46 @@ std::string FormatSubVersion(const std::string &name, int nClientVersion, const
+    {
+        LOGA("Total length of network version string (%i) exceeds maximum length (%i). Reduce "
+             "the number or size of uacomments.",
+            subver.size(), MAX_SUBVERSION_LENGTH);
```
Not sure how concerned we are about log spam, but logging every time we call `FormatSubVersion` may produce a lot of it if we need to truncate the subversion.  With the changes applied in this PR it will be called once on startup, once when the debug dialog is set up, anytime we connect to a new peer and send our version, and anytime the RPC call to `getnetworkinfo` is made.

sickpig replied:
fair point, I guess we should add a category for that logging statement. Not sure which one, thou. I'd say that NET is the best match even if it's not perfect. 

We could even imagine to remove the log message entirely since we are going to adding it back on `init.cpp` if the computed subver is equal to MAX_SUBVERSION_LENGTH as you suggested further down in the review.

I think will do the latter.

Justaphf commented on src/clientversion.cpp:
```diff
@@ -116,17 +116,46 @@ std::string FormatSubVersion(const std::string &name, int nClientVersion, const
+        int temp = 0;
+        int *ptemp = &temp;
+        if (sizeof(ptemp) == 4)
+            uacomments.push_back("32bit");
```
If we add this to the back of `uacomments`, and the user specified `uacomments` necessitate truncation due to excess length, then this is the first thing that would be lost.  I would suggest putting it at the front of `uacomments` instead so it immediately follows the EB/AD parameters and won't get truncated.

OTOH, it will force up to an extra 7 characters to be truncated from the user's specified comments, so maybe it does make more sense to put it at the very end and let it be truncated first.

Either way, I think a comment should be added to indicate why we specifically chose to put it at the front or the back of the vector.

sickpig replied:
fair points. Since I see this as a temporary thing to asses the proportion of 32bit nodes among all nodes, I would rather adding it in front of `uacomments` vector along with a code comments that explain the rationale.

Justaphf commented on src/clientversion.cpp:
```diff
@@ -116,17 +116,46 @@ std::string FormatSubVersion(const std::string &name, int nClientVersion, const
+    {
+        int temp = 0;
+        int *ptemp = &temp;
+        if (sizeof(ptemp) == 4)
```
Does it make sense to have an else case and show "64bit" as well, or do we only care about 32-bit?

sickpig replied:
we care about 32bit only, but since that they are complementary we can infer 64bit by taking the total number of nodes and subtracting the 32bit nodes count

sickpig replied:
on second thought I think Id add and `else` clause to show 64bit, so that we don't make unit tests outcome archs dependent.

Justaphf commented on src/clientversion.cpp:
```diff
@@ -116,17 +116,46 @@ std::string FormatSubVersion(const std::string &name, int nClientVersion, const
+    }
+    // If this is a 32bit build then append an identifier since we'd like to know
+    // how many still run this configuration.
+    {
```
Should inclusion of this build type information be an option the user can specifically turn on/off with a configuration setting/tweak?

I know this is something @ptschip wanted to add to get a feel for how many 32-bit nodes still exist out there in an effort to see if we can drop support for 32-bit, but it seems likely that some users wouldn't want to publicly expose even this little extra information about their node if they don't have to.

sickpig replied:
fair point. Still I see this a as a temporary measure to just guess estimate the proportion of 32bit nodes. I think that this piece of code could be removed in the release after 1.6.0.1
Having said that I'm not opposed at having a tweak.

Justaphf commented on src/clientversion.cpp:
```diff
@@ -116,17 +116,46 @@ std::string FormatSubVersion(const std::string &name, int nClientVersion, const
+    for (std::string &cmt : mapMultiArgs["-uacomment"])
+    {
+        if (cmt != SanitizeString(cmt, SAFE_CHARS_UA_COMMENT))
+            break;
```
Should we break or continue here?  If we break then we'll skip any additional comments which may be valid, but continue to create the subversion string based on any uacomments which were added prior to this failure.

NOTE: I'm not sure this `SanitizeString` check is necessary at all given that the startup routine in `init.cpp` will throw a fatal error if this same check fails there, and afaict the `mapMultiArgs["-uacomment"]` value cannot change after initialization on startup.  Though it doesn't hurt to have it either, provided we answer the break/continue question.

sickpig replied:
fair point. I'm going to remove the check from `clientversion.cpp`

Justaphf commented on src/clientversion.cpp:
```diff
@@ -116,17 +116,46 @@ std::string FormatSubVersion(const std::string &name, int nClientVersion, const
+    if (subver.size() > MAX_SUBVERSION_LENGTH)
+    {
+        LOGA("Total length of network version string (%i) exceeds maximum length (%i). Reduce "
+             "the number or size of uacomments.",
```
Log message should be updated to indicate that we truncated the comments to fit.  Previous comment expected a fatal error to be thrown.

Suggest: "Total length of network version string (%i) exceeds maximum length (%i) and has been truncated.  Reduce the number or size of uacomments to avoid truncation."

sickpig replied:
~~fair. will change.~~

removed entirely since we put it back to`init.cpp` as you described in this [comment](https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1830#discussion_r301392968)

Justaphf commented on src/clientversion.cpp:
```diff
@@ -116,17 +116,46 @@ std::string FormatSubVersion(const std::string &name, int nClientVersion, const
+        LOGA("Total length of network version string (%i) exceeds maximum length (%i). Reduce "
+             "the number or size of uacomments.",
+            subver.size(), MAX_SUBVERSION_LENGTH);
+        subver = subver.substr(0, MAX_SUBVERSION_LENGTH - 3) + ")/";
```
This is truncating 1 extra character.  Should be `MAX_SUBVERSION_LENGTH - 2`

sickpig replied:
correct, will fix.

Justaphf commented on src/init.cpp:
```diff
@@ -1337,20 +1337,15 @@ bool AppInit2(Config &config, thread_group &threadGroup)
-            strSubVersion.size(), MAX_SUBVERSION_LENGTH));
-    }
+
+    std::string strSubVersion = FormatSubVersion(CLIENT_NAME, CLIENT_VERSION, BUComments);
```
In order to still give the user immediate feedback, without requiring them to look at the log file, I think it makes sense to still have a pop-up via `InitWarning` in lieu of the `InitError` that's being replaced.  Suggest something like

```
    std::string strSubVersion = FormatSubVersion(CLIENT_NAME, CLIENT_VERSION, BUComments);
    if (strSubVersion.size() == MAX_SUBVERSION_LENGTH)
        InitWarning(strprintf(_("Total length of network version string with uacomments added exceeded "
            "the maximum length (%i) and have been truncated.  Reduce the number or size of uacomments "
            "to avoid truncation."), MAX_SUBVERSION_LENGTH));
```

NOTE: This solution isn't perfect because the user could have entered in uacomments which exactly hits the max length without the need for truncation.  To be able to determine if truncation was actually performed we'd have to modify FormatSubVersion to be able to return this information in addition to the strSubVersion.

sickpig replied:
ok this is neat. will add even if it's not "prefect" like you mentioned

Justaphf reviewed (changes requested):
@sickpig I took another pass of your changes and put in several comments/questions.  A number of them are related to the inclusion of the changes incorporated from #1827 so it's kind of a review for both features.

sickpig committed:
Remove truncation version from FormatSubVersion and it back to init.cpp

While doing that use InitWarning rather than a LOGA, since we are doing
the check during bootstrap and that uacomment can't be modified at runtime.

NOTE: This solution isn't perfect because the user could have entered in
uacomments which exactly hits the max length without the need for truncation.
To be able to determine if truncation was actually performed we'd have to
modify FormatSubVersion to be able to return this information in addition
to the strSubVersion.

sickpig committed:
Remove uacomment sanitization from FormatSubVersion()

This sanitisation is performed during the bootstrap phase (init.cpp)
and since uacomments can't be changed at runtime, the removed check
was redundant.

sickpig committed:
Add arch identifier just after EB/AD, rather than at the end of  subver.

sickpig committed:
Fix an off-by-one error while trimming subver string

sickpig committed:
Check that subver is correctly trimmed at 256 chars length

sickpig committed:
Make the unit test for FormatSubVersion to handle different architecture properly

sickpig commented:
@Justaphf PR updated according to your last round of feedback

Justaphf commented on src/clientversion.cpp:
```diff
@@ -116,17 +116,39 @@ std::string FormatSubVersion(const std::string &name, int nClientVersion, const
+
+    // If this is a 32bit build then append an identifier since we'd like to know
+    // how many still run this configuration.
+    // We are goin yo put it at the front of uacomments so it immediately
```
Nit: "goin yo" => "going to".

sickpig replied:
fixed

Justaphf commented on src/test/util_tests.cpp:
```diff
@@ -573,17 +575,43 @@ BOOST_AUTO_TEST_CASE(test_FormatParagraph)
+
+    std::string subver = FormatSubVersion("Test", 99900, BUComments);
+    BOOST_CHECK_EQUAL(subver.size(), MAX_SUBVERSION_LENGTH);
 }
```
Looking at the excessive blocks tests, I'm wondering if we should set EB, AD, and BUComments back to the defaults at the end of the test, similar to what the buip005 test case does.  Not sure if it is necessary, but since it is affecting global state could potentially cause issues with other tests run after.

sickpig replied:
Agreed, done.

Justaphf reviewed (approved):
@sickpig Two minor points, but otherwise this looks good to merge.  Since it also includes the 32/64 bit tag in the subversion string, this should also be marked as superseding #1827 and closing #1061

sickpig committed:
Reset back EB/AD to default values

gandrewstone commented:
note: add tweak to hide 32/64 bit reporting in the subver field.

sickpig committed:
Tweak to turn on/off the display of node architecture on subver string

sickpig commented:
@gandrewstone tweak added and also unit test for it
